src/Kernel/RectorKernel.php

```diff
@@ -12,54 +12,72 @@
 use Rector\Core\DependencyInjection\CompilerPass\MergeImportedRectorConfigureCallValuesCompilerPass;
 use Rector\Core\DependencyInjection\CompilerPass\RemoveSkippedRectorsCompilerPass;
 use Rector\Core\Exception\ShouldNotHappenException;
+use Rector\Core\Util\Hasher;
+use Rector\Testing\PHPUnit\StaticPHPUnitEnvironment;
 use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
 use Symfony\Component\DependencyInjection\ContainerBuilder;
 use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\DependencyInjection\Dumper\PhpDumper;
+use Symplify\SmartFileSystem\SmartFileSystem;
+use Webmozart\Assert\Assert;
 
 final class RectorKernel
 {
+    /**
+     * @var string
+     */
+    private const CACHE_KEY = 'kernel-v4';
```

Comment:
when changing dependencies of registered services without also changing existing config files - like e.g. in https://github.com/rectorphp/rector-src/pull/3817/ -, one need to increase this key to invalidate existing permanent compiled DI containers

Reply:
with [ecf3dc6](https://github.com/rectorphp/rector-src/pull/3809/commits/ecf3dc6344e4c73fc4fb2f3d9810f3c88eff9be1) I have added a new case, which automatically invalidates the compiled container cachen, when exceptions are thrown while getting services out of the container - to improve DX so we don't need to update the cache key for most operations

Reply:
Ok, I see this note